### DISCUSSION TITLE
Request for Firmware Update to Address WiFi Module Issue on Raspberry Pi Pico W

### DISCUSSION THREAD

**Discussion by mtinet** [micropython/micropython#12608] [Using MicroPython] [answered]
Subject: Request for Firmware Update to Address WiFi Module Issue on Raspberry Pi Pico W

Hello fellow makers,

I have been utilizing the Raspberry Pi Pico W for various projects and find it to be an outstanding toolkit for individuals like us. However, I, along with several peers, have stumbled upon a persistent issue with the WiFi module, which seems to malfunction during extended WiFi connections. Although the device works fine under normal conditions, this particular problem significantly affects its reliability during long-term connectivity.

Here’s how the issue manifests:

OSError: [Errno 104] ECONNRESET
OSError: [Errno 103] ECONNABORTED
Connection lost -- EOF
Use Stop/Restart to reconnect.
Process ended with exit code 1.
We have tried numerous solutions including modifying the code, executing routine reboots every 2 hours, and flashing the memory, yet none of these measures have resolved the issue.

Interestingly, it's been noted that the ESP8266 doesn’t exhibit similar problems under the same conditions. This observation leads me to believe that a firmware update for the Raspberry Pi Pico W might be a potential solution.

I am sincerely hoping for insights from this community regarding this matter. Has anyone else experienced this and found a workaround? Or is there a possibility of a firmware update that could rectify this issue and enhance the WiFi module’s stability during long-term connections?

I truly appreciate any feedback or suggestions that could help improve the user experience for many of us facing this same issue.

Thank you for your time and insights!

Warm regards,

Juhyun Kim
[mtinet@hanmail.net](mailto:mtinet@hanmail.net)

**Comment by Josverl** [comment]
Do you have a (minimal) sample/test program that can be used to replicate this issue across different wifi networks?

If so can you reference or add it here?

**Reply by mtinet** [reply]
Hi. 

This is the github address of the code. 

https://github.com/mtinet/smartFarm/tree/main/pico

main.py files is the code as occur the problem. And also another file as like main(thread).py and main(non-OLED).py has same problem.

Can you check this problem?

**Reply by mtinet** [reply]
Please check this code. The code also stopped after a few attempts. 

//code//
```python
from machine import Pin, ADC
import urequests
import time
import network

SSID = "U+Net454C"         # 공유기의 SSID를 따옴표 안에 넣으세요.
password = "DDAE014478"       # 공유기의 password를 따옴표 안에 넣으세요.

# 와이파이 연결하기
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
if not wlan.isconnected():
    # 와이파이 연결하기
    wlan.connect(SSID, password)  # 12, 13번 줄에 입력한 SSID와 password가 입력됨
    print("Waiting for Wi-Fi connection", end="...")
    print()
    while not wlan.isconnected():
        print(".", end="")
        time.sleep(1)
else:
    print(wlan.ifconfig())
    print("WiFi is Connected")
    print()
    
# Firebase Realtime Database 주소
url = "https://smartfarm-f867f-default-rtdb.firebaseio.com/"

# LED와 조도 센서를 연결할 핀 설정
led = Pin(1, Pin.OUT)  # 생장 LED 제어 핀
light = ADC(26)  # 조도 센서 연결 핀

while True:
    # 조도 센서에서 값 읽기
    light_value = int(light.read_u16()/16)
    
    # Firebase에 조도 센서 값 업데이트
    urequests.patch(url + "smartFarm.json", json={'cds': light_value})
    
    # 조도 센서 값이 1500보다 크면 LED 켜기, 아니면 LED 끄기
    if light_value > 1500:
        led.value(1)
    else:
        led.value(0)
    
    # 일정 시간 동안 대기 (예: 3초)
    time.sleep(3)
```


//result//
```python
>>> %Run -c $EDITOR_CONTENT
('192.168.123.102', '255.255.255.0', '192.168.123.1', '192.168.219.1')
WiFi is Connected

Traceback (most recent call last):
  File "<stdin>", line 37, in <module>
  File "urequests.py", line 192, in patch
  File "urequests.py", line 83, in request
OSError: [Errno 12] ENOMEM
>>> 
```

**Comment by beetlegigg** [comment, marked as answer]
Note the following from the docs:
> It's mandatory to close response objects as soon as you finished
> working with them. On MicroPython platforms without full-fledged
> OS, not doing so may lead to resource leaks and malfunction.

I think urequests.patch is a response object so maybe you need to close it:
```
response = urequests.patch......
response.close()
```
But I'm no expert in this, I just had a play with urequests some months ago.

**Reply by beetlegigg** [reply]
Woops, sorry for large text, a cut and paste gone a bit wobbly :-)

**Reply by andrewleech** [reply]
This is quite correct, the code will currently be leaking ram and connections with every while loop because the response hasn't been closed.
Hope you don't mind @beetlegigg I updated the formatting of your comment, the content was already correct :-)

**Reply by mtinet** [reply]
Thank you all. I will try the solution right now!!! Thank you very much! And I will be comment the result of your suggesstion.

**Reply by mtinet** [reply]
Unfortunately, I encountered the same issue... It stopped.

Every attempt I've made has been unsuccessful.

response.close() isn't effective.
gc.collect() has no effect.
The exception 'continue' isn't helping.
micropython.mem_info(1) isn't resolving the issue.
All my efforts lead to the system stopping within 24 hours.

How can I address this issue?

**Comment by mtinet** [comment]
Can anyone help me out!

**Reply by scruss** [reply]
You're making a request every three seconds. You only need to make a request when the status of the sensor changes

**Reply by mtinet** [reply]
Yes. I was making like that. But my main project is referencing realtime database in firebase google. Is that effective of your method? Because I want to make the realtie web controller for my main project. Anyway I'll try suggesstion of your comment. Thank you very much!!!!!!

**Reply by fdufnews** [reply]
You're making a request every three seconds, is it possible you exceed the maximum request a day?
If so maybe you can be temporary banned

**Reply by mtinet** [reply]
Ah!!! I'll check the firebase policy. Thank you!!!!
